### PR TITLE
Add a timeout for task container to become ready.

### PR DESCRIPTION
**Pull request by jeevb**

### CODE REVIEW

File: cosmos/job/drm/drm_gke.py [hunk 1]
```diff
@@ -52,6 +52,7 @@
 PREEMPTIBLE_NODE_LABEL = 'cloud.google.com/gke-preemptible'
 PREEMPTIBLE_START_GRACE_PERIOD = timedelta(hours=2)
 LONG_RUNNING_JOB_THRESHOLD = timedelta(hours=4)
+CONTAINER_READY_GRACE_PERIOD = timedelta(minutes=30)
```
Comment (shields-fn) [inline cosmos/job/drm/drm_gke.py]: Aren't containers normally ready in tens of seconds?  If it isn't ready in 3-5 minutes, I wouldn't expect it to become ready ever.

File: cosmos/job/drm/drm_gke.py [hunk 2]
```diff
@@ -154,7 +155,7 @@ class GkeDRMOptions(NamedTuple):
     memory_limit: Optional[str] = None
     # TODO (jeev): Add support for disk requirements for tasks
     # disk: Optional[str] = None
-    timeout: Optional[int] = None
+    timeout: Optional[timedelta] = None
```
Comment (shields-fn) [inline cosmos/job/drm/drm_gke.py]: Nice, I like the type safety here. (Having seen too many real bugs where people set 30,000-second timeouts when they meant to set 30,000-millisecond timeouts, or the opposite.)

File: cosmos/job/drm/drm_gke.py [hunk 3]
```diff
@@ -1146,7 +1153,33 @@ def _fetch_pod_statuses(self, *tasks):
...
+                continue
+
+            # Skip handling of Unknown status - wait for the cluster to resolve
+            # this.
+            if condition.status == 'Unknown':
+                return
+
+            # Parse the amount of time spent in current status
+            elapsed_since_last_transition = (
+                datetime.utcnow() -
```
Comment (shields-fn) [inline cosmos/job/drm/drm_gke.py]: Is the timezone conversion necessary?

File: cosmos/job/drm/drm_gke.py (hunk 1 above)
Comment (jeevb) [reply cosmos/job/drm/drm_gke.py]: Pods can be stuck in `ContainerCreating` while pulling the image as well (~3 mins in our case). Also I wanted to give the cluster a chance to resolve these issues on its own - I've seen containers take > 7 mins sometimes to start up, but do so just fine.

File: cosmos/job/drm/drm_gke.py (hunk 2 above)
Comment (jeevb) [reply cosmos/job/drm/drm_gke.py]: I got it from your last review comment. :)

File: cosmos/job/drm/drm_gke.py [hunk 4]
```diff
@@ -1206,12 +1246,13 @@ def _fetch_completed_tasks(self, *tasks):
                 wrapped_task.logger.warning(
                     f"Pod '{wrapped_task.pod_name}' associated with "
                     f"{wrapped_task} has exceeded deadline of "
-                    f"{wrapped_task.options.timeout} second(s). Timing out..."
+                    f"{wrapped_task.options.timeout}. Timing out..."
```
Comment (annadcunningham) [inline cosmos/job/drm/drm_gke.py]: will this include the unit of time on its own?

File: cosmos/job/drm/drm_gke.py (hunk 3 above)
Comment (jeevb) [reply cosmos/job/drm/drm_gke.py]: Unfortunately, yes. The kubernetes API objects have UTC timestamps with tz set, and `datetime` doesn't allow operations involving datetime+tz and datetime-tz. BUT, even if you remove the tz in the `datetime` object, you still have to make sure you are comparing compatible timestamps (UTC in this case). I tried a few things, and this is the least complicated combination that worked.

File: cosmos/job/drm/drm_gke.py (hunk 4 above)
Comment (jeevb) [reply cosmos/job/drm/drm_gke.py]: It doesn't, but its inferrable. It basically looks like `01:30:00` for an hour and 30 minutes.

File: cosmos/job/drm/drm_gke.py (hunk 3 above)
Comment (shields-fn) [reply cosmos/job/drm/drm_gke.py]: I keep forgetting how broken Python's time APIs are. There's no reason this shouldn't work. Go does the right thing: https://play.golang.org/p/Sjlf9HZonUW

File: cosmos/job/drm/drm_gke.py (hunk 1 above)
Comment (shields-fn) [reply cosmos/job/drm/drm_gke.py]: (capturing offline discussion)
That's a good reason then, although sometime we should figure out what is happening in that seven minutes, which is a crazy long time. Modern computers should be able to do a ton of work in seven minutes.

**Pull request merged**